contracts/src/CrossChainApplications/NativeTokenBridge/NativeTokenDestination.sol

```diff
@@ -0,0 +1,207 @@
+            "NativeTokenDestination: zero Source Contract Address"
+        );
+        nativeTokenSourceAddress = nativeTokenSourceAddress_;
+
+        initialReserveImbalance = initialReserveImbalance_;
+        currentReserveImbalance = initialReserveImbalance_;
+    }
+
+    /**
+     * @dev See {ITeleporterReceiver-receiveTeleporterMessage}.
+     *
+     * Receives a Teleporter message and routes to the appropriate internal function call.
+     */
+    function receiveTeleporterMessage(
+        bytes32 senderBlockchainID,
+        address senderAddress,
+        bytes calldata message
+    ) external nonReentrant {
+        // Only allow the Teleporter messenger to deliver messages.
+        require(
+            msg.sender == address(teleporterMessenger),
+            "NativeTokenDestination: Unauthorized TeleporterMessenger contract"
+        );
+
+        // Only allow messages from the source chain.
+        require(
+            senderBlockchainID == sourceBlockchainID,
+            "NativeTokenDestination: Invalid Source Chain"
+        );
+
+        // Only allow the partner contract to send messages.
+        require(
+            senderAddress == nativeTokenSourceAddress,
+            "NativeTokenDestination: Unauthorized Sender"
+        );
+
+        (address recipient, uint256 amount) = abi.decode(
+            message,
+            (address, uint256)
+        );
+        require(
+            recipient != address(0),
+            "NativeTokenDestination: zero Recipient Address"
+        );
+        require(amount != 0, "NativeTokenDestination: Transfer value of 0");
+
+        uint256 adjustedAmount = amount;
+        if (currentReserveImbalance > 0) {
+            if (amount > currentReserveImbalance) {
+                emit CollateralAdded({amount: currentReserveImbalance, remaining: 0});
+                adjustedAmount = amount - currentReserveImbalance;
+                currentReserveImbalance = 0;
+            } else {
+                currentReserveImbalance -= amount;
+                emit CollateralAdded({amount: amount, remaining: currentReserveImbalance});
+                return;
+            }
+        }
+
+        // Calls NativeMinter precompile through INativeMinter interface.
+        _nativeMinter.mintNativeCoin(recipient, adjustedAmount);
+        totalMinted += adjustedAmount;
+        emit NativeTokensMinted(recipient, adjustedAmount);
+    }
+
+    /**
+     * @dev See {INativeTokenDestination-transferToSource}.
+     */
+    function transferToSource(
+        address recipient,
+        TeleporterFeeInfo calldata feeInfo,
+        address[] calldata allowedRelayerAddresses
+    ) external payable nonReentrant {
+        // The recipient cannot be the zero address.
+        require(
+            recipient != address(0),
+            "NativeTokenDestination: zero Recipient Address"
+        );
+
+        require(
+            currentReserveImbalance == 0,
```

Comment:
Since we're doing unchecked in-place arithmetic on `currentReserveImbalance`, it may be less error prone to change this check to `currentReserveImbalance <= 0`. The logic in `receiveTeleporterMessage` looks sound to me, such that `currentReserveImbalance` could never be less than 0, but it would be very easy to accidentally introduce a bug there that would then be improperly checked here. Just my 2c.

Reply:
`currentReserveImbalance` is a `uint64`, so cannot be less than 0. We could check for underflow, but the logic would be exactly the same as what is currently in the if/else statement.